### PR TITLE
[Perf] Add skip_clone to SamplingParams for internal request handling

### PR DESCRIPTION
**Pull request by mgoin**
## Purpose

This PR adds a `skip_clone` flag to `SamplingParams` that allows skipping the expensive `copy.deepcopy()` call in `input_processor.process_inputs()` when the params object is created fresh per request and dedicated to a single request lifecycle.

The `skip_clone=True` flag is set in all `to_sampling_params()` methods in the OpenAI protocol classes, the demo API server, and beam search implementations. When set, `SamplingParams.clone()` returns self instead of performing a deep copy.

For `n>1` parallel sampling, it uses Python's `copy.copy()` for child requests rather than the internal clone() method. We might be able to remove this since logits processors aren't supported anymore, which I think was the original reason.

FIX https://github.com/vllm-project/vllm/issues/29081

cc @njhill @DarkLight1337 @NickLucche @Jialin 

## Test Plan

Will trigger full CI, and I should verify perf with a profile

## Test Result

Before:
Screenshot 2025-12-19 144852

After:
Screenshot 2025-12-19 144927

Notice the clone is completely gone. Here is the script used to highlight input overhead:

```python
"""
VLLM_TORCH_PROFILER_DIR=./profile vllm serve Qwen/Qwen3-0.6B --disable-log-requests
python bench_clone.py
"""
import asyncio
import httpx
from openai import AsyncOpenAI

URL = "http://localhost:8000"
N, CONCURRENCY = 500, 50

client = AsyncOpenAI(base_url=f"{URL}/v1", api_key="x")
http = httpx.AsyncClient()

async def bench():
    sem = asyncio.Semaphore(CONCURRENCY)
    async def req(i):
        async with sem:
            await client.completions.create(
                model="Qwen/Qwen3-0.6B", prompt=f"{i}", max_tokens=1)
    await asyncio.gather(*[req(i) for i in range(N)])

async def main():
    await bench()  # warmup
    await http.post(f"{URL}/start_profile")
    await bench()
    await http.post(f"{URL}/stop_profile")

asyncio.run(main())

```

For a quick benchmark on my local machine with Qwen3-0.6B, I saw a small throughput improvement

```
(vllm) ➜  vllm git:(main) ✗ python bench_clone.py 
Requests: 500
Time: 1.82s
Throughput: 274.5 req/s
Avg latency: 3.64ms
(vllm) ➜  vllm git:(sampling-params-skip-clone) ✗ python bench_clone.py
Requests: 500
Time: 1.76s
Throughput: 284.3 req/s
Avg latency: 3.52ms
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/entrypoints/openai/protocol.py [hunk 1]
```diff
@@ -844,6 +845,7 @@ def to_sampling_params(
             bad_words=self.bad_words,
             allowed_token_ids=self.allowed_token_ids,
             extra_args=extra_args or None,
+            skip_clone=True,  # Created fresh per request, safe to skip clone
```
[the remark this answers is not in the document]
Comment (mgoin) [reply vllm/entrypoints/openai/protocol.py:880]: @njhill did you say before that we can just remove these logits processors? I'm not sure how it works with the custom registered ones now in v1

[the remark this answers is not in the document]
Comment (njhill) [reply vllm/entrypoints/openai/protocol.py:880]: @mgoin yes, I think we can remove that field now

[the remark this answers is not in the document]
Comment (mgoin) [reply vllm/entrypoints/openai/protocol.py:880]: Okay I'll remove logits_processors in a follow up to keep the changes separable

**Comment by mgoin** [comment]
@DarkLight1337 I found that SamplingParams does get mutated in the EngineCore, mostly just for harmless things like output kind. However this makes it clear to me that we should be shallow copying at the moment. We can explore frozen params later on but I fear it will require a larger refactor. The shallow copy takes care of the bulk of the performance issue

**Comment by njhill** [comment]
Thanks @mgoin! I had actually been thinking more about the best way to handle this, to not require the explicit use of the new field. I think this is a great change for now but we might want to make it clear that this new field is currently for internal use only and may be removed in a future update, I can open a PR to adjust the comment for that...
